project-docs/stumptown-writers-guide.md

```diff
@@ -0,0 +1,697 @@
+
+```
+
+The `title`, `mdn_url`, `specifications`, and `recipe` properties are all the same as in a normal recipe-driven page like `html-element`. The `related_content` is in the landing page itself rather than in the recipe: this is because different landing pages want different sidebars, while all the HTML elements should get the same sidebar.
+
+The other property is a `link_lists` object describing the lists of links to include.
+
+The Markdown part of the landing page provides an overview of the technology.
+
+## Guide pages
+
+For reference pages (and to a lesser extent landing pages) we've described a system where a page has a clear structure that's described in the recipe and enforced by stumptown.
+
+I'd estimate that about 90% of the web documentation pages on MDN are reference pages and will be (we hope) amenable to this treatment.
+
+But MDN also contains a substantial number of "guide" pages. These pages are fundamentally unstructured prose content, and the recipe model doesn't work for them. Still, we can't treat them just as prose, because they want to embed special content such as browser compatibility tables or live samples - and they want to be able to embed them just anywhere.
+
+In stumptown, guide pages are Markdown files with YAML front matter. The front matter is quite minimal:
+
+```yaml
+---
+title: Applying color to HTML elements using CSS
+mdn_url: /en-US/docs/Web/HTML/Applying_color
+related_content: /content/related_content/html.yaml
+recipe: guide
+---
+
+```
+
+Like everything else, it has a `title` and a `mdn_url` and a `recipe`. Like landing pages, it has `related_content` in the front matter rather than the recipe, because because different guide pages want different sidebars. These are the only items allowed in the front matter.
+
+For guide pages, the recipe is empty, so doesn't put any constraints on the things that the page can contain.
+
+The Markdown part is of course the guide's prose. To support the "special content" that guides want to embed, stumptown has the concept of "directives". A directive looks like:
+
+```
+{{{{{embed-example:examples/specifying-colors-in-stylesheets}}}}}
+```
+
+We currently support two directives:
+
+* "embed-example": this tells stumptown that the guide wants to embed a live sample at this point. Its argument is "examples/specifying-colors-in-stylesheets", which is a path to a directory containing an example. The content of this directory is just the same as the directories references by the `examples` front matter property, with a mandatory "description.md" and optional "example.[html|css|js]" files.
+
+* "embed-compat": this tells stumptown that the guide wants to embed a BCD table at this point. Its argument is a BCD query string like "html.elements.audio".
+
+We will probably add support for more directives in future, not too many. One likely candidate would be "embed-link-list".
+
+Directives can be inserted anywhere in guide pages. They can only be used in guide pages.
```

Comment:
Why only guide pages? I ask because I strongly suspect we will come to regret that limitation eventually. Especially if it winds up being the mechanism used to embed media files and the like.

Reply:
Firstly because it makes everything more complex. Secondly, because it's part of the concept of reference pages that they can be quite formally structured, and letting people put random things like this wherever they like undermines this.

Mostly though, because it really breaks the idea that the content can be easily machine-accessed. At the moment in stumptown, a section of prose is just a section of relatively simple HTML. So it's possible, for example, for a consumer to fetch just the "Accessibility concerns" section of the doc, and display it in the context of an accessibility inspector. But what if this section contains an "embed-compat" directive? Well:
* the JSON output could include the directive, like our ?raw endpoint. Then the consumer has to understand how to process directives.
* the JSON output could include the rendered BCD table. Then stumptown-content is determining how the content ought to be rendered, which we really want to avoid.
* the JSON output could break up "Accessibility concerns" into the prose pieces and the interleaved BCD. This is what we do for guide pages. But then there isn't an "Accessibility concerns" at all.
* ?

> I ask because I strongly suspect we will come to regret that limitation eventually.

One advantage stumptown has over Kuma is that Kuma had to anticipate what the likely use cases were going to be. That's really hard and means you necessarily end up building lots of things you don't really need (but then find it hard to remove for various reasons). But stumptown is dealing with a really mature content base. So we know, pretty much, what the use cases are. It doesn't seem very likely that people will start writing radically new types of JS docs at this point.